src/v/redpanda/application.cc

```diff
@@ -559,6 +559,24 @@ void application::check_environment() {
     storage::directories::initialize(
       config::node().data_directory().as_sstring())
       .get();
+
+    if (config::shard_local_cfg().storage_strict_data_init()) {
+        // Look for the special file that indicates a user intends
+        // for the found data directory to be the one we use.
+        auto strict_data_dir_file
+          = config::node().strict_data_dir_file_path().string();
+        auto file_exists = ss::file_exists(strict_data_dir_file).get();
+
+        if (!file_exists) {
+            vlog(
+              _log.error,
+              "Data directory not in expected state: {} not found, is the "
+              "expected filesystem mounted?",
+              strict_data_dir_file);
+            throw std::invalid_argument(
```

Comment:
optional: this might be a good time for a quick refactor to the exception type used in syschecks: currently it's using runtime_error, this is using invalid_argument, both should really be using some explicit type (syscheck_failed?), that way `application::run` can catch that type separately to the generic exception handler, and log the message without an ugly std::something prefix.

Reply:
Should it be a separate refactoring PR? If not I'll add it into this one.